### PR TITLE
Add GET multi-site forecast route + many little cleanups

### LINKED ISSUE #43: get multi sites on routes
## Detailed Description
Would be great to be able to get mutlie sites from the following routes
- /sites/pv_actual/{site_uuid}
- /sites/pv_forecast/{site_uuid}

## Context
UI will do multiple calls to the UI

## Possible Implementation
make an option of adding mulite sites in a list for both these sites, and then returning a list of these objects

### PR DESCRIPTION
**Pull request by simlmx**
_See descriptions in individual commits_

Fixes #43 

### Some notes on the query for the forecast

We used to return the latest forecast value for each distinct (site, forecast.timestamp_utc) pair, which I think is very unintuitive and somewhat broken.

For instance, if we were to run the forecast at a different interval than the space between our different horizons (which is the case currently although it might be by accident), then we'll get something really funny. For instance if we run inference every 5 minutes, but predict for 15 minute brackets, we'll get interlaced predictions for 3 different forecast `timestamp_utc`. This is not what we assume we are seeing when looking at a forecast chart. It could also lead to some funny patterns. If we want the 5 minute resolution, we can always interpolate the 15 minute one, or change or brackets.

Another funny behaviour of the old query is that for anything after "now", we'll get proper forecasts (if we ignore the previously mentioned problem), but for the timestamps *before* "now", we'll get the `[0, 15min]` predictions, always, from all the different forecasts. If we want to show predictions that we made in the past, it's probably best to pick some horizon (I think we use 4 hours for the national forecast?), and make another query for that.

In the new version I have here, we simply return the latest forecast for each site, which I think is what we are intuitively expecting when we look at the forecast charts.

### CODE REVIEW

**Comment by peterdudfield** [comment]
> _See descriptions in individual commits_
> 
> Fixes #43
> 
> ### Some notes on the query for the forecast
> We used to return the latest forecast value for each distinct (site, forecast.timestamp_utc) pair, which I think is very unintuitive and somewhat broken.
> 
> For instance, if we were to run the forecast at a different interval than the space between our different horizons (which is the case currently although it might be by accident), then we'll get something really funny. For instance if we run inference every 5 minutes, but predict for 15 minute brackets, we'll get interlaced predictions for 3 different forecast `timestamp_utc`. This is not what we assume we are seeing when looking at a forecast chart. It could also lead to some funny patterns. If we want the 5 minute resolution, we can always interpolate the 15 minute one, or change or brackets.
> 
> Another funny behaviour of the old query is that for anything after "now", we'll get proper forecasts (if we ignore the previously mentioned problem), but for the timestamps _before_ "now", we'll get the `[0, 15min]` predictions, always, from all the different forecasts. If we want to show predictions that we made in the past, it's probably best to pick some horizon (I think we use 4 hours for the national forecast?), and make another query for that.
> 
> In the new version I have here, we simply return the latest forecast for each site, which I think is what we are intuitively expecting when we look at the forecast charts.

Thanks @simlmx  - I like the changes.

I agree there weill be some issues if we dont forecast at the same interval. I would rather pass than down the line for the moment / make a github issue for it. 

On changing the route names, I actaully prefer the current way we do it. I can see why you have done it. This is becasuse our other API has `national` and `gsp` as the main folder, so `sites` makes sense. For some reason I prefer the `uuid` at the end, perhaps it makes it more readble when seeing what route has been hit. If you were to change this back, I would be all good for merging this

**Review by peterdudfield — CHANGES_REQUESTED** [review]
see comment on main. 

Also could we add a test for getting multiple sites?

**Comment by braddf** [comment]
Hey, throwing my two pence in here:

For the routes, I do think these:
`/sites`
`/sites/pv_actual`
`/sites/pv_forecast`
`/sites/<site_uuid>/pv_actual`
`/sites/<site_uuid>/pv_forecast`
are more RESTful than what we currently have, REST resource wise - is there any reason we are doing this the other way around with the `<uuid>` not after the resource it's referring to in the route @peterdudfield?

I also said to Simon yesterday that we would before long have an extra authorisation layer needed to be able to check/retrieve the specific sites of a client, without relying on query params... which maybe doesn't directly inform the route names, but keeping them more standard I think would be better before we get any more complicated..?

**Comment by peterdudfield** [comment]
> Hey, throwing my two pence in here:
> 
> For the routes, I do think these: `/sites` `/sites/pv_actual` `/sites/pv_forecast` `/sites/<site_uuid>/pv_actual` `/sites/<site_uuid>/pv_forecast` are more RESTful than what we currently have, REST resource wise - is there any reason we are doing this the other way around with the `<uuid>` not after the resource it's referring to in the route @peterdudfield?
> 
> I also said to Simon yesterday that we would before long have an extra authorisation layer needed to be able to check/retrieve the specific sites of a client, without relying on query params... which maybe doesn't directly inform the route names, but keeping them more standard I think would be better before we get any more complicated..?

Thats good then if its more RESTful, then go for it

Just to adress a point above. I think the pv_forecast should displaye
- from now on, the lastest forecast made
- from before now, the last forecast value made for each target interval
- Later on we can think about showing a different time horizon

File: pv_site_api/main.py [hunk 1]
```diff
@@ -243,25 +277,81 @@ async def get_pv_actual(site_uuid: str, session: Session = Depends(get_session))
...
+def _get_latest_forecast_by_sites(session: Session, site_uuids: list[str]) -> list[Forecast]:
+    """Get the latest forecast for given site uuids.
+
+    Return:
+    ------
+        Pydantic models representing the data.
+    """
+    # Get the latest forecast for each site.
+    subquery = (
+        session.query(ForecastSQL)
```
Comment (peterdudfield) [inline pv_site_api/main.py:304]: please use the read function in the pv-datamodel

Comment (peterdudfield) [reply pv_site_api/main.py:304]: This is what it used to be
```
generations_sql = get_pv_generation_by_sites(
        session=session, start_utc=start_utc, site_uuids=[uuid.UUID(site_uuid)]
```

If you want to update it then please update in the datamodel

Comment (peterdudfield) [reply pv_site_api/main.py:304]:
```  
 latest_forecast_values = get_latest_forecast_values_by_site(
        session=session, site_uuids=[site_uuid], start_utc=start_utc
    )
```

was what used to be there

**Comment by simlmx** [comment]
I pushed some fixes in a last commit: c8811b22910606aa6f515e01ffc191452daf5475
Thanks for reviewing those too!
@braddf @peterdudfield

**Pull request merged**